cube/generic_networks/att_model.py

```diff
@@ -0,0 +1,215 @@
+import dynet as dy
+from tqdm import tqdm
+import numpy as np
+from numpy import mean, argmax
+from get_data import get_data
+from numpy import ceil
+import sys
+
+
+"""
+	Attention mechanism based sequence classification model.
+	Inspired from https://nlp.stanford.edu/pubs/emnlp15_attn.pdf.
+"""
+
+class AttentionClassifier(object):
+	def __init__(self, num_of_classes,
+	 vocab_size,
+	 embedding_size=100,
+	 lstm_num_of_layers=1, 
+	 state_size=200, 
+	 batch_size=2):
+
+		self.embedding_size = embedding_size
+		self.lstm_num_of_layers = lstm_num_of_layers
+		self.num_of_classes = num_of_classes
+		self.batch_size = batch_size
+		self.state_size = state_size
+		self.vocab_size = vocab_size + 1
+
+	def fit(self, X_train=None, y_train=None, X_test=None, y_test=None, epochs=None):
```

Comment:
Please split this method and use a batch learning controlled from the parameters passed in main.py, namely --batch-size.

Also add a separate trainer for this, and display useful information such as:
- percentage during training (multiples of 5%)
- loss on the training set (after each epoch)
- epoch duration (execution time=...)
- a relevant metric on the development set after evaluation

Please store your models as follows:
 - use .last to indicate a checkpoint of the model after every epoch
 - use .best<Metric> to indicate a checkpoint for the best performing model on the development set.

I'm pasting here an excerpt  for you to use as an example:

```text
Starting training for parser
==PARAMETERS==
TRAIN FILE: ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-train.conllu
DEV FILE: ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-dev.conllu
EMBEDDINGS FILE: ../corpus/embeddings/wiki.dummy.vec
STOPPING CONDITION: 20
OUTPUT BASE: ../models/UD_English_SML/parser
AUX SOFTMAX WEIGHT: 0.2
CONFIG FILE: None
==============

Reading ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-train.conllu... found 12543 sequences
Reading ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-dev.conllu... found 2002 sequences
Computing encoding maps... done
Unique words: 17966
Unique chars: 75
Unique labels: 50
Unique UPOS: 17
Unique XPOS: 50
Holistic word count: 2954
200 303
Orthogonal pretrainer loss: 1.08e-26
200 303
Orthogonal pretrainer loss: 8.78e-27
300 400
Orthogonal pretrainer loss: 3.85e-23
300 400
Orthogonal pretrainer loss: 3.57e-23
300 900
Orthogonal pretrainer loss: 1.89e-30
300 900
Orthogonal pretrainer loss: 1.90e-30
200 800
Orthogonal pretrainer loss: 1.01e-30
200 800
Orthogonal pretrainer loss: 1.21e-30
200 600
Orthogonal pretrainer loss: 9.44e-31
200 600
Orthogonal pretrainer loss: 9.27e-31
200 600
Orthogonal pretrainer loss: 1.01e-30
200 600
Orthogonal pretrainer loss: 1.02e-30
Storing encodings in ../models/UD_English_SML/parser.encodings
Storing config in ../models/UD_English_SML/parser.conf
Starting epoch 1
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=89.07018480937295 execution_time=1129.7310800552368
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.5248131064100525 LAS=0.3065850167011293 UPOS=0.648043581994592 XPOS=0.530658501670113 ATTRS=0.58827739780$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 2
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=48.41013286851823 execution_time=1141.804716348648
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.6983060283123906 LAS=0.5644186416414825 UPOS=0.8104024176872913 XPOS=0.7141323365675203 ATTRS=0.749045649$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 3
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=34.46338212515743 execution_time=1136.4327113628387
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.7422061396532528 LAS=0.6536106250994115 UPOS=0.8678622554477493 XPOS=0.8225703833306823 ATTRS=0.843009384$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 4
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=28.836603429417647 execution_time=1133.9607968330383
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.7812549705742007 LAS=0.7145697470971847 UPOS=0.8923970097025609 XPOS=0.8686575473198664 ATTRS=0.889852075$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 5
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=25.650525833710162 execution_time=1138.0560524463654
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.8002226817241928 LAS=0.7442738985207571 UPOS=0.90726896771115 XPOS=0.8896134881501511 ATTRS=0.90929696198$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 6
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=23.388270108882086 execution_time=1138.4073255062103
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.8153729918880229 LAS=0.7618896134881501 UPOS=0.9118021313822173 XPOS=0.8984014633370447 ATTRS=0.915460473$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
```